### ISSUE TITLE
Windows launcher doesn't quote the executable

### ISSUE THREAD

**Issue by eryksun** [pypa/setuptools#622]
In `run` on [line 305 in launcher.c](https://github.com/pypa/setuptools/blob/v23.1.0/launcher.c#L305), the quoted executable path is added to the args array (i.e. `*newargsp++ = quoted(ptr)`), but this is subsequently ignored by  `join_executable_and_args`. Instead it uses the unquoted executable path. Thus if Python is installed to "C:\Program Files", this will mistakenly execute "C:\Program.exe" if it exists. 

For example, here are the `executable` and `args` parameters passed to `join_executable_and_args`: 

    0:000> ?? (char *)@rcx
    char * 0x000001d6`aefbe100
     "C:\Program Files\Python35\python.exe"

    0:000> ?? ((char **)@rdx)[0]
    char * 0x000001d6`aefb4140
     ""C:\Program Files\Python35\python.exe""
    0:000> ?? ((char **)@rdx)[1]
    char * 0x000001d6`aefb9aa0
     ""C:\Program Files\Python35\Scripts\hello-script.py""

And here is the resulting command line passed to `CreateProcessA`:

    Breakpoint 1 hit
    KERNELBASE!CreateProcessA:
    00007ffb`9ef0ca00 4c8bdc          mov     r11,rsp
    0:000> ?? (char *)@rdx
    char * 0x000001d6`aefb6a30
     "C:\Program Files\Python35\python.exe "C:\Program Files\Python35\Scripts\hello-script.py""

`join_executable_and_args` should just be `join_args` since the quoted executable is already the first item in `args`.

I'd prefer that setuptools abandoned this old launcher code to instead use the [simple launcher](https://bitbucket.org/vinay.sajip/simple_launcher) from distlib, which supports Unicode, a job object, and embedded scripts. pip uses distlib's launcher for entry points in wheel packages.

**Comment by anthrotype** [comment]
> I'd prefer that setuptools abandoned this old launcher code to instead use the simple launcher from distlib

I was working on that, but then got busy with other stuff.
See discussion at: https://github.com/pypa/setuptools/pull/595
I hope to find some time to finish this next week -- unless you want to give it go :)

**Comment by anthrotype** [comment]
this issue looks like a duplicate of https://github.com/pypa/setuptools/issues/398.

@fkrull is working on a PR to fix that: https://github.com/pypa/setuptools/pull/616

**Comment by eryksun** [comment]
No, it's not the same problem. In that issue, the lack of quoting in the shebang itself causes problems with the [`_splitpath`](https://msdn.microsoft.com/en-us/library/e737s6tf.aspx) call. You end up with a defunct command line like the following:

    0:000> ?? (char *)@rdx
    char * 0x00000000`007c53d0
     "C:\Program "Files\Python35\python.exe" "C:\Program Files\Python35\Scripts\hello-script.py""

This command will *only* execute "C:\Program.exe". With the bug that I'm reporting, it executes "C:\Program.exe" *if it exists*, but otherwise will execute "C:\Program Files\Python35\python.exe" because... Windows.

You seem to have things well in hand in #595. What do you think about creating a common pypa package to share vendored packages between setuptools and pip?

**Comment by jaraco** [comment]
Also, there's #592, in which the test for this is failing. So at some point, this was working and has since regressed, but since our continuous integration doesn't include Windows and I no longer develop on Windows as a matter of course, that platform gets second-class treatment.

**Comment by jaraco** [comment]
We now have continuous integration for Windows at https://ci.appveyor.com/project/jaraco/setuptools
